sktime/annotation/eagglo.py

```diff
@@ -0,0 +1,430 @@
+        self.progression[0, :] = [
+            sum(self.sizes[:i]) if i > 0 else 0 for i in range(self.n_cluster + 1)
+        ]  # N + 1 for cyclic mergers
+
+        # array to specify the starting point of a cluster
+        self.lm = np.zeros(2 * self.n_cluster - 1, dtype=int)
+        self.lm[: self.n_cluster] = range(self.n_cluster)
+
+    def _gof_update(self, i: int) -> float:
+        """Compute the updated goodness-of-fit statistic, left cluster given by i."""
+        fit = self.gof_[-1]
+        j = self.right[i]
+
+        # get new left and right clusters
+        rr = self.right[j]
+        ll = self.left[i]
+
+        # remove unneeded values in the gof_
+        fit -= 2 * (
+            self.distances[i, j] + self.distances[i, ll] + self.distances[j, rr]
+        )
+
+        # get cluster sizes
+        n1 = self.sizes[i]
+        n2 = self.sizes[j]
+
+        # add distance to new left cluster
+        n3 = self.sizes[ll]
+        k = (
+            (n1 + n3) * self.distances[i, ll]
+            + (n2 + n3) * self.distances[j, ll]
+            - n3 * self.distances[i, j]
+        ) / (n1 + n2 + n3)
+        fit += 2 * k
+
+        # add distance to new right
+        n3 = self.sizes[rr]
+        k = (
+            (n1 + n3) * self.distances[i, rr]
+            + (n2 + n3) * self.distances[j, rr]
+            - n3 * self.distances[i, j]
+        ) / (n1 + n2 + n3)
+        fit += 2 * k
+
+        return fit
+
+    def _find_closest(self, K: int) -> Tuple[int, int]:
+        """Determine which clusters will be merged_, for K clusters.
+
+        Greedily optimize the goodness-of-fit statistic by merging the pair of adjacent
+        clusters that results in the largest increase of the statistic's value.
+
+        Parameters
+        ----------
+        K: int
+            Number of clusters
+
+        Returns
+        -------
+        result : Tuple[int, int]
+            Tuple of left cluster and right cluster index values
+        """
+        best_fit = -1e10
+        result = (0, 0)
+
+        # iterate through each cluster to see how the gof_ value changes if merged_
+        for i in range(K + 1):
+            if self.open[i]:
+                gof_ = self._gof_update(i)
+                if gof_ > best_fit:
+                    best_fit = gof_
+                    result = (i, self.right[i])
+
+        self.gof_ = np.append(self.gof_, best_fit)
+        return result
+
+    def _update_distances(self, i: int, j: int, K: int) -> None:
+        """Update distance from new cluster to other clusters, store to self."""
+        # which clusters were merged_, info only
+        self.merged_[K - self.n_cluster + 1, 0] = (
+            -i if i <= self.n_cluster else i - self.n_cluster
+        )
+        self.merged_[K - self.n_cluster + 1, 1] = (
+            -j if j <= self.n_cluster else j - self.n_cluster
+        )
+
+        # update left and right neighbors
+        ll = self.left[i]
```

Comment:
Personally I think it might be cleaner to have a function which takes care of getting the left/right neighbors rather than maintaining two separate arrays as such. Not necessary change though - just have seen approaches like this lead to unexpected errors.  Easier when there is one set of ground truth and other values are extracted in relation to that.

Reply:
Can you elaborate on what unexpected errors might happen and under what circumstances?

Reply:
Ah - I think it is fine as is!  This was not a change request.  I was more thinking that you can derive a lot of these arrays (eg left/right, open etc) by looking at relationships to the current cluster array. There isn't anything wrong with keeping them as separate array though, so no need to change it. All I meant is that once you have a bunch of redundancy it can be easy to have errors sprout up if you are not on top of maintaining that redundancy everywhere.  Please consider this resolved though, was more of a future suggestion.

Reply:
ie was more thinking - in the future writing this differently might reduce bugs.

Reply:
I agree 100% 👍